### ISSUE TITLE
[TorchDispatch] Scalar Only Inputs Gets Matched To Tensor Schema

### ISSUE THREAD

**Issue by eellison** [pytorch/pytorch#86128]
### 🐛 Describe the bug

When you call `torch.div` with scalar-only inputs, the schema that gets matched to is `aten::div.Tensor(Tensor self, Tensor other) -> Tensor`.  

```
import torch
from torch.utils._python_dispatch import TorchDispatchMode

class DebugMode(TorchDispatchMode):
    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        print(func, args, kwargs)
        return func(*args, **kwargs)

with DebugMode():
    print(torch.div(240, 8))
```
> aten.div.Tensor (240, 8) {}
tensor(30.)

I would expect the scalar inputs to get promoted to Tensors before the `aten.div` is invoked, or for a scalar schema to get matched to (probably the former). 

Also, I couldn't find the tests for scalar only inputs to Binary torch functions. 

Related: https://github.com/pytorch/torchdynamo/issues/1445

### Versions

master

**Comment by ezyang** [comment]
This makes sense to me. Scalar override doesn't affect Python argument parsing. We accept numbers by wrapping numbers into tensors. Therefore the Tensor overload gets hit.

**Comment by eellison** [comment]
The numbers aren't being wrapped by the point of torch dispatch. The arguments are `(240, 8)`. If i'm invoking a schema with arguments `(Tensor, Tensor)`, I would expect to find tensors there, not scalars.

**Comment by ezyang** [comment]
But that's now how wrapped number works. A wrapped number always manifests in Python as a number, never as a tensor, because a "wrapped number tensor" is not an actual thing in the Python UI.

Yes, this is confusing, but it's the intended semantics. If you want to force this to go to a Scalar, Scalar overload, that's going to be a bit of refactoring on how Python argument parsing works.

**Comment by eellison** [comment]
Why would it not work to wrap them to 0 element tensors ?

**Comment by ezyang** [comment]
Semantics are different. 

```
>>> torch.zeros(2, dtype=torch.int32) * torch.tensor(2, dtype=torch.double)
tensor([0., 0.], dtype=torch.float64)
>>> torch.zeros(2, dtype=torch.int32) * 2.0
tensor([0., 0.])
```

**Comment by eellison** [comment]
It's only different because you're instantiating a tensor of size 2 (not a 0-dim tensor). I'm talking abt wrapping to 0-dim when all the arguments are scalars.

**Comment by eellison** [comment]
I think the reason for the difference there is just because you instatiated a tensor with `torch.double` instead of `torch.float`.

```
>>> torch.zeros(2, dtype=torch.int32) * torch.tensor(2, dtype=torch.float)
tensor([0., 0.])
>>> torch.zeros(2, dtype=torch.int32) * 2.0
tensor([0., 0.])
```

**Comment by eellison** [comment]
We could decompose `aten.add.Tensor` and other ops that allow scalars as inputs to preserve the semantics. We also match to `aten.add.Tensor` instead of `aten.add.Scalar` even when the output dtype is the same, currently.

**Comment by eellison** [comment]
cc @bobrenjc93 @ezyang could the recent tensorify work help here ?

**Comment by ezyang** [comment]
Sort of? If the inputs were float and then eventually got to a torch.div, tensorify would turn them into tensors and make the tensor inputs. But I don't recall why this was a problem for you in the first place, and there are lots of similar situations (e.g., inputs are int) where tensorify doesn't trigger right now.
